File: kakao.app/js/ked/view/menu/macro.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.prototype.hasOwnProperty(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}}

var macro

import kxk from "../../../kxk.js"
let post = kxk.post

import belt from "../../edit/tool/belt.js"

import theme from "../../theme/theme.js"

import inputchoice from "./inputchoice.js"


macro = (function ()
{
    _k_.extend(macro, inputchoice)
    function macro (screen)
    {
        this.screen = screen
    
        this["hide"] = this["hide"].bind(this)
        this["arrange"] = this["arrange"].bind(this)
        macro.__super__.constructor.call(this,this.screen,'macro')
        this.isVisible = false
    }

    macro.prototype["arrange"] = function ()
    {
        var c, h, ih, iz, scx, scy, w, x, y

        w = this.width + 2 + 1
        c = this.choices.numChoices()
        ih = (this.inputIsActive() ? 2 : 0)
        iz = _k_.max(0,ih - 1)
        h = c + 2 + ih
        scy = parseInt(this.screen.rows / 2)
        y = parseInt(scy - (c + 2) / 2)
        y -= ih
        scx = parseInt(this.screen.cols / 2)
        x = parseInt(scx - w / 2)
        this.input.layout(x + 2,y + 1,w - 4,iz)
        this.choices.layout(x + 1,y + 1 + ih,w - 2,c)
        return this.cells.layout(x,y,w,h)
    }

    macro.prototype["show"] = function ()
    {
        var ccol, items

        items = belt.linesForText(`diff
commit`)
        items = items.map(function (i)
        {
            return ' ' + i
        })
        ccol = parseInt(this.screen.cols / 2) - 5
        this.width = belt.widthOfLines(items)
        this.input.set('')
        this.input.hide()
        this.choices.set(items)
        this.choices.select(0)
        this.choices.state.setView([0,0])
        return macro.__super__.show.call(this)
    }

    macro.prototype["hide"] = function ()
    {
        post.emit('focus','editor')
        return macro.__super__.hide.call(this)
    }

    macro.prototype["applyChoice"] = function (choice)
    {
        var input, num

        if (_k_.empty(choice))
        {
            input = this.input.current()
            if (input === 'bof')
            {
                post.emit('goto.bof')
            }
            else if (input === 'eof')
            {
                post.emit('goto.eof')
            }
            else if (num = parseInt(input))
            {
                post.emit('goto.line',num - 1,'ind')
            }
        }
        this.hide()
        return true
    }

    return macro
})()

export default macro;

File: kakao.app/love/lua/view/menu/macro.lua
--[[
    00     00   0000000    0000000  00000000    0000000   
    000   000  000   000  000       000   000  000   000  
    000000000  000000000  000       0000000    000   000  
    000 0 000  000   000  000       000   000  000   000  
    000   000  000   000   0000000  000   000   0000000   

    menu for actions that don't have keyboard shortcuts
--]]

inputchoice = require "view.menu.inputchoice"


local macros = class("macros", inputchoice)
    


function macros:init() 
        inputchoice.init(self, 'macro')
        return self
    end

--  0000000   00000000   00000000    0000000   000   000   0000000   00000000  
-- 000   000  000   000  000   000  000   000  0000  000  000        000       
-- 000000000  0000000    0000000    000000000  000 0 000  000  0000  0000000   
-- 000   000  000   000  000   000  000   000  000  0000  000   000  000       
-- 000   000  000   000  000   000  000   000  000   000   0000000   00000000  


function macros:arrange() 
        local w = max(20, ((self.width + 2) + 1)) -- 2 for frame, 1 for right 
        local c = self.choices:numChoices()
        
        local ih = (function () 
    if self:inputIsActive() then 
    return 2 else 
    return 0
             end
end)()
        local iz = max(0, (ih - 1))
        
        local h = ((c + 2) + ih)
        local scy = int((_G.screen.rows / 2))
        local y = int((scy - ((c + 2) / 2)))
        y = y - ih
        
        local scx = int((_G.screen.cols / 2))
        local x = int((scx - (w / 2)))
        
        self.input:layout((x + 2), (y + 1), (w - 4), iz)
        self.choices:layout((x + 1), ((y + 1) + ih), (w - 2), c)
        return self.cells:layout(x, y, w, h)
    end

--  0000000  000   000   0000000   000   000
-- 000       000   000  000   000  000 0 000
-- 0000000   000000000  000   000  000000000
--      000  000   000  000   000  000   000
-- 0000000   000   000   0000000   00     00


function macros:show() 
        local items = belt.linesForText
        
            [[
status
diff
history
]]
        
        items = items:map(function (i) 
    return ' ' .. i
end) -- padding for  
        
        local ccol = (int((_G.screen.cols / 2)) - 5)
        
        self.width = belt.widthOfLines(items)
        
        self.input:set('')
        self.input:hide()
        self.choices:set(items)
        self.choices:select(0)
        self.choices.state:setView(array(0, 0))
        
        return inputchoice.show(self)
    end


function macros:hide() 
        post:emit('focus', 'editor')
        return inputchoice.hide(self)
    end

--  0000000   00000000   00000000   000      000   000  
-- 000   000  000   000  000   000  000       000 000   
-- 000000000  00000000   00000000   000        00000    
-- 000   000  000        000        000         000     
-- 000   000  000        000        0000000     000     


function macros:applyChoice(choice) 
        -- log "macro.applyChoice ▸#{choice}◂"
        
        if empty(choice) then 
            local input = self.input:current()
            if (input == 'bof') then post:emit('goto.bof')
            elseif (input == 'eof') then post:emit('goto.eof')
            elseif tonumber(input) then post:emit('goto.line', tonumber(input), 'ind')
            elseif kstr.startsWith(input, 'vibrant ') then 
                    local sp = input:split(" ")
                    local sf = tonumber(sp[1])
                    local lf = (function () 
    if (#sp > 2) then 
    return tonumber(sp[2]) else 
    return 1
                         end
end)()
                    post:emit('theme.vibrant', sf, lf)
            elseif input.startsWith('class ') then 
                    local sp = input:split(" ")
                    if valid(sp[1]) then 
                        post:emit('file.class', sp[1])
                    end
            end
        else 
            if (choice == 'diff') then post:emit('differ.file')
            elseif (choice == 'status') then post:emit('differ.status')
            elseif (choice == 'history') then post:emit('differ.history')
            end
        end
        
        self:hide()
        return true
    end

return macros